libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,993 @@
+      memcpy((char*)data_buffer + pos, kt.data(), kt.size());
+      pos += kt.size();
+   });
+
+   key_type s((const char*)data_buffer, data_size);
+
+   if (data_size > detail::max_stack_buffer_size) {
+      free(data_buffer);
+   }
+   return s;
+}
+/* @endcond */
+
+// This is the "best" way to document a function that does not technically exist using Doxygen.
+#if EOSIO_CDT_DOXYGEN
+/**
+ * @brief A function for converting types to the appropriate binary representation for the EOSIO Key Value database.
+ * @details The CDT provides implementations of this function for many of the common primitives and for structs/tuples.
+ * If sticking with standard types, contract developers should not need to interact with this function.
+ * If doing something more advanced, contract developers may need to provide their own implementation for a special type.
+ */
+template <typename T>
+inline key_type make_key(T val) {
+   return {};
+}
+#endif
+
+
+/**
+ * Used to return the appropriate representation of a case insensitive string for the EOSIO Key Value database.
+ * This is only valid for ASCII strings.
+ *
+ * @param val - The string to be made case-insensitive
+ * @return The binary representation of the case-insensitive string
+ */
+inline key_type make_insensitive(const std::string& val) {
+   return make_key(val, true);
+}
+
+
+/**
+ * @defgroup keyvalue Key Value Table
+ * @ingroup contracts
+ *
+ * @brief Defines an EOSIO Key Value Table
+ * @details EOSIO Key Value API provides a C++ interface to the EOSIO Key Value database.
+ * Key Value Tables require 1 primary index, of any type that can be serialized to a binary representation.
+ * Key Value Tables support 0 or more secondary index, of any type that can be serialized to a binary representation.
+ * Indexes must be a member variable or a member function.
+ *
+ * @tparam T         - the type of the data stored as the value of the table
+  */
+template<typename T>
+class kv_table {
+
+   enum class kv_it_stat {
+      iterator_ok     = 0,  // Iterator is positioned at a key-value pair
+      iterator_erased = -1, // The key-value pair that the iterator used to be positioned at was erased
+      iterator_end    = -2, // Iterator is out-of-bounds
+   };
+
+   struct index_config {
+      uint64_t db_name;
+      eosio::name contract_name;
+      eosio::name index_name;
+      eosio::name primary_index_name;
+      key_type prefix;
+   };
+
+   class iterator {
+   public:
+      iterator(uint32_t itr, kv_it_stat itr_stat, const index_config& config) : itr{itr}, itr_stat{itr_stat}, config{config} {}
+
+      iterator(iterator&& other) :
+         itr(std::exchange(other.itr, 0)),
+         itr_stat(std::move(other.itr_stat)),
+         config(std::move(other.config))
+      {}
+
+      ~iterator() {
+         if (itr) {
+            internal_use_do_not_use::kv_it_destroy(itr);
+         }
+      }
+
+      iterator& operator=(iterator&& other) {
+         itr = std::exchange(other.itr, itr);
```

Comment:
We need the old iterator to get cleaned up (`kv_it_destroy`) which the destructor does, so my thought was by using exchange the old iterator would get cleaned up properly.

Reply:
It's technically safe, but it's generally better to clean up resources sooner rather than later.

Reply:
In that case, would the recommended solution be to call `kv_it_destroy` on the old itr here in the move equals and then not bother with the exchange?

Reply:
Yes, I think it would be best to call `kv_it_destroy` on the old value of `itr`, set `itr` to `other.itr` and then set `other.itr` to 0 to ensure that it doesn't call `kv_it_destroy` on an already destroyed iterator when the other value's destructor gets called.

Reply:
I also think it makes sense to set `other.itr_stat` to `status::iterator_end` both here and in the move constructor. That way the other is always in a consistent state (even though the user really shouldn't be doing anything with that move-from iterator except either calling the destructor or assigning to it).